include/deal.II/base/exceptions.h

```diff
@@ -1659,11 +1659,22 @@ namespace deal_II_exceptions
 #    endif /*ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST*/
 #  endif   /*KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST*/
 #else      /*ifdef DEBUG*/
-#  define Assert(cond, exc) \
-    do                      \
-      {                     \
-      }                     \
+/*
+ * In order to avoid unused parameters (etc.) warnings we need to use cond
+ * and exc without actually evaluating the expression and generating code.
+ * We accomplish this by using decltype(...) and create a dummy pointer
+ * with these signatures.
+ */
+#  define Assert(cond, exc)                                                    \
+    do                                                                         \
+      {                                                                        \
+        typename std::remove_reference<decltype(cond)>::type *deal_ii_asser_a; \
+        typename std::remove_reference<decltype(exc)>::type  *deal_ii_asser_b; \
```

Comment:
That does really count as a "use" to compilers?

Reply:
To GCC at least, yes. This has an unused variable warning for `a_ptr`, not `a`:
```cpp
int main()
{
    int a;

    decltype(a) *a_ptr;
}
```
so
```cpp
int main()
{
    int a;

    decltype(a) *a_ptr;
    (void)a_ptr;
}
```
fixes it.